state-chain/pallets/cf-funding/src/lib.rs

```diff
@@ -769,6 +713,114 @@ impl<T: Config> Pallet<T> {
+				);
+			}
+		}
+
+		if let Some(bound_address) = BoundRedeemAddress::<T>::get(&account_id) {
+			ensure!(
+				bound_address == address || restricted_balances.contains_key(&address),
+				Error::<T>::AccountBindingRestrictionViolated
+			);
+		}
+
+		// In case the balance is lower than the sum of restricted addresses we take this
+		// discrepancy into account so that restricted addresses can still redeem.
+		let restricted_deficit: FlipBalance<T> = restricted_balances
+			.values()
+			.copied()
+			.sum::<FlipBalance<T>>()
+			.saturating_sub(T::Flip::balance(&account_id));
+
+		// The available funds are the total balance minus whichever is larger from:
+		// - The bond.
+		// - The total restricted funds that need to remain in the account after the redemption.
+		let liquid_balance = T::Flip::balance(&account_id).saturating_sub(max(
+			T::Flip::bond(&account_id),
+			restricted_balances.values().copied().sum::<FlipBalance<T>>().saturating_sub(
+				restricted_deficit + restricted_balances.get(&address).copied().unwrap_or_default(),
+			),
+		));
+
+		let redemption_fee = match amount {
+			RedemptionAmount::Max if liquid_balance == T::Flip::balance(&account_id) =>
+				Zero::zero(),
+			_ => RedemptionTax::<T>::get(),
+		};
+
+		let (debit_amount, redeem_amount) = match amount {
+			RedemptionAmount::Max =>
+				(liquid_balance, liquid_balance.saturating_sub(redemption_fee)),
+			RedemptionAmount::Exact(amount) => (amount.saturating_add(redemption_fee), amount),
+		};
+
+		ensure!(
+			T::Flip::try_burn_fee(&account_id, redemption_fee).is_ok(),
+			Error::<T>::InsufficientBalance
+		);
+
+		let mut total_restricted_balance: FlipBalance<T> = T::Amount::zero();
+
+		// If necessary, update account restrictions.
+		if let Some(restricted_balance) = restricted_balances.get_mut(&address) {
+			// Use the full debit amount here - fees are paid by restricted funds by default.
+			total_restricted_balance = *restricted_balance;
+			restricted_balance.saturating_reduce(debit_amount);
+			// ensure that the remaining restricted balance is zero or above MinimumFunding
+			ensure!(
+				restricted_balance.is_zero() || *restricted_balance >= MinimumFunding::<T>::get(),
+				Error::<T>::RestrictedBalanceBelowMinimumFunding
+			);
+
+			if restricted_balance.is_zero() {
+				restricted_balances.remove(&address);
+			}
+			RestrictedBalances::<T>::insert(&account_id, &restricted_balances);
+		}
+
+		let remaining_balance = T::Flip::balance(&account_id)
+			.checked_sub(&redeem_amount)
+			.ok_or(Error::<T>::InsufficientBalance)?;
+
+		ensure!(
+			remaining_balance == Zero::zero() || remaining_balance >= MinimumFunding::<T>::get(),
+			Error::<T>::BelowMinimumFunding
+		);
+		ensure!(
+			remaining_balance >=
+				restricted_balances.values().copied().sum::<FlipBalance<T>>() -
+					restricted_deficit,
+			Error::<T>::InsufficientUnrestrictedFunds
+		);
+
+		Ok((total_restricted_balance.saturating_sub(redemption_fee), redeem_amount))
```

Comment:
What is the purpose of returning `total_restricted_balance.saturating_sub(redemption_fee)` here? 

Couldn't we return the actual restricted amount:

```
min(
    total_restricted_balance.saturating_sub(redemption_fee),
    redeem_amount,
)
```

Then we can use this in the redeem extrinsic and also in the transfer extrinsic.